[RLlib][Docs] Added RLModule user-guide to the docs (#33909)
ray-project/ray: doc/source/rllib/rllib-rlmodule.rst

         - |pytorch| |tensorflow|
         - |pytorch|
         -
         - |pytorch|
       * - **APPO**
         - |tensorflow|
         - |tensorflow|
         - 
         - 
         - 

RLModule is a neural network container that implements three public methods: :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_train`, :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration`, and :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference`. Each method corresponds to a distinct reinforcement learning phase.

:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` handles acting and data collection, balancing exploration and exploitation. On the other hand, the :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference` serves the learned model during evaluation, often being less stochastic.

:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_train` manages the training phase, handling calculations exclusive to computing losses, such as learning Q values in a DQN model.

Enabling RL Modules in the Configuration
-----------------------------------------

> You can enable RLModules by setting the ``_enable_rl_module_api`` flag to ``True`` in the configuration object. 

Reviewer: ```suggestion
Enable RLModules by setting the ``_enable_rl_module_api`` flag to ``True`` in the configuration object. 
```

.. code-block:: python

    import torch
    from pprint import pprint

    from ray.rllib.algorithms.ppo import PPOConfig

    config = (
        PPOConfig()
        .framework("torch")
        .environment("CartPole-v1")
        .rl_module(_enable_rl_module_api=True)
    )

    algorithm = config.build()

    # run for 2 training steps
    for _ in range(2):
        result = algorithm.train()
        pprint(result)

Constructing RL Modules
-----------------------

> RLModule API provides a unified way to define custom reinforcement learning models in RLlib. This API enables users to design and implement their own models to suit their specific needs.

Reviewer: ```suggestion
RLModule API provides a unified way to define custom reinforcement learning models in RLlib. This API enables users to design and implement their own models to suit their specific needs.
```
```suggestion
RLModule API provides a unified way to define custom reinforcement learning models in RLlib. This API enables you to design and implement your own models to suit specific needs.
```

.. tabbed:: Multi Agent

    .. code-block:: python

        import gymnasium as gym
        from ray.rllib.core.rl_module.rl_module import SingleAgentRLModuleSpec
        from ray.rllib.core.rl_module.marl_module import MultiAgentRLModuleSpec
        from ray.rllib.core.testing.torch.bc_module import DiscreteBCTorchModule
        from ray.rllib.core.testing.bc_algorithm import BCConfigTest
        from ray.rllib.examples.env.multi_agent import MultiAgentCartPole

        config = (
            BCConfigTest()
            .environment(MultiAgentCartPole, env_config={"num_agents": 2})
            .rl_module(
                _enable_rl_module_api=True,
                rl_module_spec=MultiAgentRLModuleSpec(
                    module_specs=SingleAgentRLModuleSpec(module_class=DiscreteBCTorchModule)
                )
            )
            .training(model={"fcnet_hiddens": [32, 32]})
        )

Writing Custom Single Agent RL Modules 
--------------------------------------

> If your algorithm is single agent algorithm (e.g. PPO, DQN) or it is an independent multi-agent algorithm (e.g. PPO-MultiAgent), you should :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`. However, in more advanced multi-agent use-cases where there is a shared communication between agents, you should extend the :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule` class. 

Reviewer: ```suggestion
For single-agent algorithms (e.g., PPO, DQN) or independent multi-agent algorithms (e.g., PPO-MultiAgent), use :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`. For more advanced multi-agent use cases with a shared communication between agents, extend the :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule` class. 
```

> From RLlib’s point of view even single agent modules will be treated as a special case of :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule` with only one module. We can create the multi-agent representation of all RLModules by calling :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.as_multi_agent`. For example:

Reviewer: ```suggestion
RLlib treats single-agent modules as a special case of :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule` with only one module. Create the multi-agent representation of all RLModules by calling :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.as_multi_agent`. For example:
```

        import tensorflow as tf

        class DiscreteBCTfModule(TfRLModule):

            def __init__(self, config: RLModuleConfig) -> None:
                super().__init__(config)

                input_dim = self.config.observation_space.shape[0]
                hidden_dim = self.config.model_config_dict["fcnet_hiddens"][0]
                output_dim = self.config.action_space.n

                self.policy = tf.keras.Sequential(
                    [
                        tf.keras.layers.Dense(hidden_dim, activation="relu"),
                        tf.keras.layers.Dense(output_dim),
                    ]
                )

                self.input_dim = input_dim

            def _forward_inference(self, batch: NestedDict) -> Mapping[str, Any]:
                return self._forward_train(batch)

            def _forward_exploration(self, batch: NestedDict) -> Mapping[str, Any]:
                return self._forward_train(batch)

            def _forward_train(self, batch: NestedDict) -> Mapping[str, Any]:
                action_logits = self.policy(batch["obs"])
                return {"action_dist": tf.distributions.Categorical(logits=action_logits)}

> One thing that can be enforced in :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` is checking the existence of certain input, output keys in the data that gets communicated into and out of RLModules. This will serve multiple purposes: 

Reviewer: ```suggestion
In :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` you can enforce the checking for the existence of certain input or output keys in the data that is communicated into and out of RLModules. This serves multiple purposes: 
```

> - It is self-documenting on what is the IO requirement of each method.

Reviewer: ```suggestion
- It is self-documenting on what is the I/O requirement of each method.
```

> - It fails fast. i.e. if users extend the modules and implement something that does not match the assumptions regarding the IO specs, it will inform them about missing keys and their expected format. For example, we always expect this RLModule to have a ``obs`` key on the input batch and also output ``action_dist`` key as the output. 

Reviewer: ```suggestion
- For failures to happen quickly. If users extend the modules and implement something that does not match the assumptions of the I/O specs, the check reports missing keys and their expected format. For example, we always RLModule should have an ``obs`` key on the input batch and an output ``action_dist`` key as the output. 
```

.. tabbed:: Type Spec

    .. code-block:: python

        from ray.rllib.models.specs.typing import SpecType
        import torch
        
        class DiscreteBCTorchModule(TorchRLModule):
            ...
            def output_specs_exploration(self) -> SpecType:
                # Enforce that output nested dict from exploration method has a key
                # "action_dist" and its value is a torch.distribution.Categorical
                return {"action_dist": torch.distribution.Categorical}

:py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` has two methods for each forward method, totaling 6 methods that can be override to describe the specs of the input and output of each method:

- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.input_specs_inference`
- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.output_specs_inference`
- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.input_specs_exploration`
- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.output_specs_exploration`
- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.input_specs_train`
- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.output_specs_train`

> To learn more about this see the documentation of `SpecType`.

Reviewer: ```suggestion
To learn more, see the `SpecType` documentation.
```

Reviewer: Suggest linking to the `SpecType` doc.

Author: It should be written in a separate PR. But temporary should be ok.

> Writing Custom Multi Agent RL Modules (Advanced)

Reviewer: ```suggestion
Writing Custom Multi-Agent RL Modules (Advanced)
```

------------------------------------------------

> For multi-agent modules, RLlib implements :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule`. It is dictionary of :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`s, one for each policy, plus possibly some shared modules. The base-class implementation works for most of use-cases where we need to defined independent neural for sub-groups of agents. For more complex, multi-agent use-cases, where the agents share some part of their neural network, one should inherit from this class and override the default implementation.

Reviewer: ```suggestion
For multi-agent modules, RLlib implements :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule`, which is a dictionary of :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`s, one for each policy, plus possibly some shared modules. The base-class implementation works for most of use-cases where we need to defined independent neural for sub-groups of agents. For more complex, multi-agent use-cases, where the agents share some part of their neural network, one should inherit from this class and override the default implementation.
```
```suggestion
For multi-agent modules, RLlib implements :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule`. It is dictionary of :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`s, one for each policy, and possibly some shared modules. The base-class implementation works for most use cases that need to define independent neural for sub-groups of agents. For more complex, multi-agent use cases, where the agents share some part of their neural network, inherit from this class and override the default implementation.
```

Reviewer: I wasn't sure who "one" was. Is that the user? or the the agent? I took a guess that it's the user.

The :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule` offers an API for constructing custom models tailored to specific needs. The key method for this customization is :py:meth:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModule`.build.

> The following example shows how we can create a custom multi-agent RL module where the underlying modules share an encoder that gets applied to the global part of the observations space, while the local part is passed through a separate encoder specific to each policy. 

Reviewer: ```suggestion
The following example creates a custom multi-agent RL module with underlying modules. The modules share an encoder, which gets applied to the global part of the observations space. The local part passes through a separate encoder specific to each policy. 
```

Reviewer: I might've mangled the meaning, but I'm trying to make it active voice.

> .. tabbed:: Multi agent with shared encoder (torch)

Reviewer: ```suggestion
.. tabbed:: Multi agent with shared encoder (Torch)
```

    .. code-block:: python

        from ray.rllib.core.rl_module.torch.torch_rl_module import TorchRLModule
        from ray.rllib.core.rl_module.marl_module import (
            MultiAgentRLModuleConfig, MultiAgentRLModule
        )
        from ray.rllib.utils.nested_dict import NestedDict

        import torch
        import torch.nn as nn

        class BCTorchRLModuleWithSharedGlobalEncoder(TorchRLModule):

>             """A RLModule with a shared encoder between agents for global observation."""

Reviewer: ```suggestion
            """An RLModule with a shared encoder between agents for global observation."""
```

        class BCTorchMultiAgentModuleWithSharedEncoder(MultiAgentRLModule):
            def __init__(self, config: MultiAgentRLModuleConfig) -> None:
                super().__init__(config)

            def build(self):

                module_specs = self.config.modules
                module_spec = next(iter(module_specs.values()))
                global_dim = module_spec.observation_space["global"].shape[0]
                hidden_dim = module_spec.model_config_dict["fcnet_hiddens"][0]
                shared_encoder = nn.Sequential(
                    nn.Linear(global_dim, hidden_dim),
                    nn.ReLU(),
                    nn.Linear(hidden_dim, hidden_dim),
                )

                rl_modules = {}
                for module_id, module_spec in module_specs.items():
                    rl_modules[module_id] = BCTorchRLModuleWithSharedGlobalEncoder(
                        encoder=shared_encoder,
                        local_dim=module_spec.observation_space["local"].shape[0],
                        hidden_dim=hidden_dim,
                        action_dim=module_spec.action_space.n,
                    )

                self._rl_modules = rl_modules

> To construct this custom multi-agent RL odule, we need to pass the class to the :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModuleSpec` constructor. We need to also pass the :py:class:`~ray.rllib.core.rl_module.rl_module.SingleAgentRLModuleSpec` for each agent. This is because RLlib needs to know the observation, action spaces, and model hyper-parameters for each agent.

Reviewer: ```suggestion
To construct this custom multi-agent RL module, pass the class to the :py:class:`~ray.rllib.core.rl_module.marl_module.MultiAgentRLModuleSpec` constructor. Also, pass the :py:class:`~ray.rllib.core.rl_module.rl_module.SingleAgentRLModuleSpec` for each agent because RLlib requires the observation, action spaces, and model hyper-parameters for each agent.
```

.. code-block:: python

    import gymnasium as gym
    from ray.rllib.core.rl_module.rl_module import SingleAgentRLModuleSpec
    from ray.rllib.core.rl_module.marl_module import MultiAgentRLModuleSpec

    spec = MultiAgentRLModuleSpec(
        marl_module_class=BCTorchMultiAgentModuleWithSharedEncoder,
        module_specs={
            "local_2d": SingleAgentRLModuleSpec(
                observation_space=gym.spaces.Dict({
                    "global": gym.spaces.Box(low=-1, high=1, shape=(2,)),
                    "local": gym.spaces.Box(low=-1, high=1, shape=(2,)),
                }),
                action_space=gym.spaces.Discrete(2),
                model_config_dict={"fcnet_hiddens": [64]},
            ),
            "local_5d": SingleAgentRLModuleSpec(
                observation_space=gym.spaces.Dict({
                    "global": gym.spaces.Box(low=-1, high=1, shape=(2,)),
                    "local": gym.spaces.Box(low=-1, high=1, shape=(5,)),
                }),
                action_space=gym.spaces.Discrete(5),
                model_config_dict={"fcnet_hiddens": [64]},
            ),
        },
    )

    module = spec.build()

> Extending Existing RLlib RL Modules

Reviewer: ```suggestion
Extending Existing RLlib RLModules
```

Reviewer: Not sure if there is a distinction between `RL Modules` and `RLModules`. In case there wasn't, I was trying to make the title consistent.

-----------------------------------

> RLlib provides a number of RLModules for different frameworks (e.g. PyTorch, TensorFlow, etc.). These modules can be extended by inheriting from them and overriding the methods you need to customize. For example, you can extend :py:class:`~ray.rllib.algorithms.ppo.torch.ppo_torch_rl_module.PPOTorchRLModule` and augment it with your own customization. You can then pass the new customized class into the algorithm configuration. 

Reviewer: ```suggestion
RLlib provides a number of RLModules for different frameworks (e.g., PyTorch, TensorFlow, etc.). Extend these modules by inheriting from them and overriding the methods you need to customize. For example, extend :py:class:`~ray.rllib.algorithms.ppo.torch.ppo_torch_rl_module.PPOTorchRLModule` and augment it with your own customization. Then pass the new customized class into the algorithm configuration. 
```

> There are two ways possible to extend existing RLModules:

Reviewer: ```suggestion
There are two possible ways to extend existing RLModules:
```

.. tabbed:: Inheriting existing RLModules

>     One way to extend existing RLModules is to inherit from them and override the methods you need to customize. For example, you can extend :py:class:`~ray.rllib.algorithms.ppo.torch.ppo_torch_rl_module.PPOTorchRLModule` and augment it with your own customization. You can then pass the new customized class into the algorithm configuration to use the PPO algorithm to optimize your custom RLModule.

Reviewer: ```suggestion
    One way to extend existing RLModules is to inherit from them and override the methods you need to customize. For example, extend :py:class:`~ray.rllib.algorithms.ppo.torch.ppo_torch_rl_module.PPOTorchRLModule` and augment it with your own customization. Then pass the new customized class into the algorithm configuration to use the PPO algorithm to optimize your custom RLModule.
```

    .. code-block:: python

        class MyPPORLModule(PPORLModule):
    
            def __init__(self, config: RLModuleConfig):
                super().__init__(config)
                ...

        # Pass in the custom RLModule class to the spec
        algo_config = algo_config.rl_module(
            rl_module_spec=SingleAgentRLModuleSpec(module_class=MyPPORLModule)
        )

    
.. tabbed:: Extending RLModule Catalog

>     Another way you can customize your module by extending their catalog. The catalog is a component that defines the default architecture and behavior of a model based on factors such as ``observation_space``, ``action_space``, and so on. To modify sub-components of an existing RLModule, extend the corresponding catalog class.

Reviewer: ```suggestion
    Another way to customize your module is by extending its catalog. The catalog is a component that defines the default architecture and behavior of a model based on factors such as ``observation_space``, ``action_space``, etc. To modify sub-components of an existing RLModule, extend the corresponding catalog class.
```

>     For instance, if you want to adapt the existing ``PPORLModule`` for a custom graph observation space not supported by RLlib out-of-the-box, you can extend the catalog class used to create the ``PPORLModule`` and override the method responsible for returning the encoder component. This will ensure your custom encoder replaces the default one initially provided by RLlib. For more information on the catalog class, refer to the `RLModule Catalog` user-guide.

Reviewer: ```suggestion
    For instance, to adapt the existing ``PPORLModule`` for a custom graph observation space not supported by RLlib out-of-the-box, extend the catalog class used to create the ``PPORLModule`` and override the method responsible for returning the encoder component to ensure that your custom encoder replaces the default one initially provided by RLlib. For more information on the catalog class, refer to the `RLModule Catalog` user guide.
```

       * - **APPO**
         - |tensorflow|
         - |tensorflow|
         - 
         - 
         - 

RLModule is a neural network container that implements three public methods: :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_train`, :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration`, and :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference`. Each method corresponds to a distinct reinforcement learning phase.

:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` handles acting and data collection, balancing exploration and exploitation. On the other hand, the :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference` serves the learned model during evaluation, often being less stochastic.

:py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_train` manages the training phase, handling calculations exclusive to computing losses, such as learning Q values in a DQN model.

Enabling RL Modules in the Configuration
-----------------------------------------

Enable RLModules by setting the ``_enable_rl_module_api`` flag to ``True`` in the configuration object. 

> .. code-block:: python

Reviewer 2: like Richard mentioned, let's make all these `literalinclude` statements. Otherwise these code blocks risk being outdated quickly. RLlib already has a [`doc_code` folder](https://github.com/ray-project/ray/tree/master/doc/source/rllib/doc_code) for that purpose. Alternatively, use an example or test folder in `rllib`, but make sure it's all tested. Thanks!
